agent/qrexec-agent.c

```diff
@@ -868,28 +868,51 @@ int fill_fds_for_select(fd_set * rdset, fd_set * wrset)
 void handle_trigger_io()
 {
     struct msg_header hdr;
-    struct trigger_service_params params;
-    int ret;
+    struct trigger_service_params3 params;
+    char *command = NULL;
+    size_t command_len;
     int client_fd;
 
     client_fd = do_accept(trigger_fd);
     if (client_fd < 0)
         return;
-    hdr.len = sizeof(params);
-    ret = read(client_fd, &params, sizeof(params));
-    if (ret == sizeof(params)) {
-        hdr.type = MSG_TRIGGER_SERVICE;
-        snprintf(params.request_id.ident, sizeof(params.request_id), "SOCKET%d", client_fd);
-        if (libvchan_send(ctrl_vchan, &hdr, sizeof(hdr)) < 0)
-            handle_vchan_error("write hdr");
-        if (libvchan_send(ctrl_vchan, &params, sizeof(params)) < 0)
-            handle_vchan_error("write params");
-    }
-    if (ret <= 0) {
-        close(client_fd);
+    if (!read_all(client_fd, &hdr, sizeof(hdr)))
+        goto error;
+    if (hdr.type != MSG_TRIGGER_SERVICE3 ||
+            hdr.len < sizeof(params) ||
+            hdr.len > MAX_SERVICE_NAME_LEN) {
```

Comment:
I think you want `hdr.len > sizeof(params) + MAX_SERVICE_LEN` here. This would also match the code in `sanitize_message_from_agent`.